### PR TITLE
Allow indexing URIs

### PR DESCRIPTION
**Pull request by omad**
### Reason for this pull request

It's impossible to index datasets from URLs and S3 buckets without writing custom code. This update allows running `datacube dataset add <URL>` where `<URL>` can refer to HTTP/HTTPS/S3 protocols, with room for extension to new protocols.

As a side effect, I've also done some refactoring to break the `utils` code into separate modules, 

### Proposed changes

- Add support for indexing URLs
- Refactor of `datacube.utils` code into separate modules
- Remove deprecated functions from `Datacube` API (`product_sources`/`product_data`/`product_observations`)
- Add docstrings to methods in `utils` and elsewhere
- Remove some Python 2 compatible code
- Expand tests to cover the new URL support

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

File: datacube/api/core.py [hunk 1]
```diff
@@ -652,6 +637,20 @@ def get_bounds(datasets, crs):
     return geometry.box(left, bottom, right, top, crs=crs)
 
 
+def set_resampling_method(measurements, resampling=None):
```
Comment (Kirill888) [inline datacube/api/core.py]: Few concerns with this one:

1. sometimes copies other times returns reference
2. not clear from the name that this is not done by reference, and you have to use the output
3. why can't I set resampling to None exactly?

File: datacube/index/hl.py [hunk 2]
```diff
@@ -192,10 +193,37 @@ def resolve_ds(ds, sources, cache=None):
...
+    (including lineage datasets). if not supplied we will consider all
+    products.
+
+    :param list exclude_products: List of products to exclude from matching
+
+    :param fail_on_missing_lineage: If True fail resolve if any lineage
+    datasets are missing from the DB
+
+    :param verify_lineage: If True check that lineage datasets in the
+    supplied document are identical to dB versions
```
Comment (Kirill888) [inline datacube/index/hl.py]: `dB` vs `DB`

File: datacube/scripts/product.py [hunk 3]
```diff
@@ -129,85 +120,84 @@ def _write_csv(index):
                                          'spatial_dimensions'], extrasaction='ignore')
     writer.writeheader()
 
-    def add_first_name(row):
-        names_ = row['name']
-        row['name'] = names_ if names_ else None
-        return row
+    # def add_first_name(row):
```
Comment (Kirill888) [inline datacube/scripts/product.py]: delete?

File: datacube/ui/expression.py [hunk 4]
```diff
@@ -1,18 +1,38 @@
...
+    eg.
+
+      - low < field < high
+      - high > field > low
+      - field in range(low, high)
+
+API
+---
+
+This module exposes and single function :func:`parse_expression` which takes
```
Comment (Kirill888) [inline datacube/ui/expression.py]: `and` -> `a`

**Review by Kirill888 — CHANGES_REQUESTED** [review]

File: datacube/model/utils.py [hunk 5]
```diff
@@ -14,7 +14,8 @@
 
 import datacube
 from datacube.model import Dataset
-from datacube.utils import geometry, SimpleDocNav, sorted_items, InvalidDocException
+from datacube.utils import geometry, SimpleDocNav, InvalidDocException
+from datacube.utils.py import sorted_items
```
Comment (uchchwhash) [inline datacube/model/utils.py]: is there a module named `py` or am I misreading this?

File: datacube/scripts/product.py (hunk 3 above)
Comment (uchchwhash) [reply datacube/scripts/product.py]: :+1:

**Review by uchchwhash — DISMISSED** [review]
Looks really good, especially the `utils` re-org.

**Review by uchchwhash — APPROVED** [review]
Looks good to me! :+1:

**Review by Kirill888 — APPROVED** [review]
good purge, and utils was overdue for splitting.

**Pull request merged**